seqr/views/apis/anvil_workspace_api_tests.py

```diff
@@ -405,32 +477,144 @@ def test_create_project_from_workspace(self, mock_tempfile, mock_mv_file, mock_s
-        mock_mv_file.assert_called_with(
+        self.mock_tempfile.assert_called_with(mode='wb', delete=False)
+        self.mock_tempfile.return_value.__enter__.return_value.write.assert_called_with(b's\nHG00735\nNA19675\nNA19678')
+        self.mock_mv_file.assert_called_with(
             TEMP_PATH, 'gs://seqr-datasets/v02/GRCh38/AnVIL_WES/{guid}/base/{guid}_ids.txt'.format(guid=project.guid),
             user=self.manager_user
         )
 
+        self._assert_dag_trigger_requests(UPDATED_ANVIL_VARIABLES, 3, project.guid, test_add_data=False)
+
+        self._assert_sent_slack_message(project, 'GRCh38')
+
+        # Test project exist
+        url = reverse(create_project_from_workspace, args=[TEST_WORKSPACE_NAMESPACE, TEST_NO_PROJECT_WORKSPACE_NAME])
+        response = self.client.post(url)
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.reason_phrase, 'Project "{name}" for workspace "{namespace}/{name}" exists.'
+                         .format(namespace=TEST_WORKSPACE_NAMESPACE, name=TEST_NO_PROJECT_WORKSPACE_NAME))
+
+        url = reverse(create_project_from_workspace, args=[TEST_WORKSPACE_NAMESPACE, TEST_NO_PROJECT_WORKSPACE_NAME2])
+        self._test_mv_file_and_triggering_dag_exception(
+            url, {'workspace_namespace': TEST_WORKSPACE_NAMESPACE, 'workspace_name': TEST_NO_PROJECT_WORKSPACE_NAME2},
+            ['HG00735', 'NA19675', 'NA19678'], 'GRCh38', REQUEST_BODY)
+
+        # Test logged in locally
+        remove_token(self.manager_user)  # The user will look like having logged in locally after the access token is removed
+        response = self.client.post(url)
+        self.assertEqual(response.status_code, 302)
+        self.assertEqual(response.url, '/login/google-oauth2?next=/api/create_project_from_workspace/my-seqr-billing/anvil-no-project-workspace2/submit')
+
+    @responses.activate
+    def test_add_workspace_data(self):
+        # Test insufficient Anvil workspace permission
+        url = reverse(add_workspace_data, args=[PROJECT2_GUID])
+        self.check_collaborator_login(url, login_redirect_url='/login/google-oauth2')
+        response = self.client.post(url, content_type='application/json', data=json.dumps({}))
+        self.assertEqual(response.status_code, 403)
+        self.assertEqual(response.json()['error'], 'Permission Denied')
+
+        # Test requesting to load data from a workspace without an existing project
+        url = reverse(add_workspace_data, args=['no_PROJECT1_GUID'])
+        response = self.client.post(url, content_type='application/json', data=json.dumps({}))
+        self.assertEqual(response.status_code, 404)
+        self.assertEqual(response.json()['error'], 'Project matching query does not exist.')
+
+        # Test missing required fields in the request body
+        url = reverse(add_workspace_data, args=[PROJECT1_GUID])
+        self.client.force_login(self.manager_user)
```

Comment:
why are we using `check_collaborator_login` but requiring a manager user here? shouldn't we just use `check_manager_login` instead?

Reply:
Okay, I didn't notice that `check_manager_login` also tests the collaborator doesn't have sufficient permissions.